src/Features/Lsif/Generator/Logging/LsifFormatLogger.cs

```diff
@@ -0,0 +1,58 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.IO;
+using Microsoft.CodeAnalysis.LanguageServerIndexFormat.Generator.Writing;
+using Microsoft.Extensions.Logging;
+using Newtonsoft.Json;
+
+namespace Microsoft.CodeAnalysis.LanguageServerIndexFormat.Generator.Logging
+{
+    internal sealed class LsifFormatLogger : ILogger
+    {
+        private readonly TextWriter _writer;
+        private readonly object _writerGate = new object();
+
+        public LsifFormatLogger(TextWriter writer)
+        {
+            _writer = writer;
+        }
+
+        public IDisposable BeginScope<TState>(TState state)
+        {
+            throw new NotImplementedException();
+        }
+
+        public bool IsEnabled(LogLevel logLevel)
+        {
+            return logLevel >= LogLevel.Information;
+        }
+
+        public void Log<TState>(LogLevel logLevel, EventId eventId, TState state, Exception? exception, Func<TState, Exception?, string> formatter)
+        {
+            var message = formatter(state, exception);
+
+            // The only supported severities are "Info", "Warning", and "Error", so we'll map critical to Error.
+            var severity = logLevel switch { LogLevel.Information => "Info", LogLevel.Warning => "Warning", _ => "Error" };
+
+            // If we have an exception, we'll use the exception message, otherwise the logged message. It doesn't appear we can log both at once.
+            var command = new CommandWithParameters("log",
+                new LogCommandParameters(severity, exception?.Message ?? message, exception?.GetType().ToString(), exception?.StackTrace));
+            var serializedCommand = JsonConvert.SerializeObject(command, LineModeLsifJsonWriter.SerializerSettings);
+
+            lock (_writerGate)
+            {
+                _writer.Write(serializedCommand);
```

Comment:
Is there enough being logged that synchronous writes + lock have any meaningful performance impact?

Reply:
At this point, no, we don't log enough for that to be a problem.